[USMON-1385] usm: sharedlibraries: Use fexit

guyarb opened the pull request:
### What does this PR do?
This PR introduces several improvements and enhancements to USM's handling of shared libraries and probes:
1. Refactors helper functions into general-purpose helpers for better reusability.
2. Adds fexit probes for `openat2` to improve performance by replacing existing tracepoints.
3. Introduces bypassable programs for fexit probes to support dynamic control of probe execution.
4. Updates the `sysOpenAt2Supported` logic to avoid kernel version-based assumptions.
5. Adds a dynamic method to determine if fexit is supported on the system.
6. Implements logic to decide which probes (fexit or tracepoints) to enable based on runtime capabilities.

### Motivation
The changes aim to:
- Improve the performance of `open` syscalls by reducing the overhead of existing tracepoints.
- Enable dynamic and efficient determination of probe compatibility with the underlying system.
- Introduce a more flexible and reusable codebase for future enhancements and testing.

### Describe how you validated your changes

- Refactored helper functions were validated through the existing unit tests ensuring correct functionality.
- Added fexit probes were benchmarked to verify the ~50% reduction in performance impact for `open` syscalls.
- Deploying on entei cluster for dogfooding

Fexit benchmarks - 
Used [bpftop](https://github.com/Netflix/bpftop) to track performance of existing and new probes, and used the following testing program to generate extreme load of `open` syscalls

```go
package main

import (
	"os"
	"os/signal"
	"sync"
	"syscall"
	"strconv"
)

func main() {
	const X = 10
	var wg sync.WaitGroup
	stop := make(chan struct{})

	for i := 0; i < X; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			fileName := "temp_file_" + strconv.Itoa(id) + ".txt"
			file, err := os.Create(fileName)
			if err != nil {
				panic(err)
			}
			file.Close()
			defer func() {
				os.Remove(fileName)
			}()
			for {
				select {
				case <-stop:
					return
				default:
					file, err := os.Open(fileName)
					if err == nil {
						file.Close()
					}
				}
			}
		}(i)
	}

	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	<-c
	close(stop)
	wg.Wait()
}
```

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

guyarb committed:
usm: sharedlibraries: Refactor helper functions

A preliminary change, to refactor 'do_sys_open_helper_enter' and 'do_sys_open_helper_exit` helper functions
into a general purpose helpers 'fill_lib_path' and 'push_event_if_relevant'. The original helpers
'do_sys_open_helper_enter' and 'do_sys_open_helper_exit` now uses 'fill_lib_path' and 'push_event_if_relevant'.

In the next commit we will use 'fiil_lib_path' and 'push_event_if_relevant' to implement fexit probe

guyarb committed:
usm: sharedlibraries: Add fexit probes for openat2

Adding fexit version for openat2 probe, to replace (decided on runtime) the 2 existing
tracpoints. Instead of having a tracpoint for 'sys_enter_openat2' and 'sys_exit_openat2',
we can use fexit on openat2, to get both input arguments and return code in a single
probe, rather then caching the input arguments via a tracepoint on 'sys_enter_openat2' and
pops them, and fetch the return code in the tracpoint on 'sys_exit_openat2'.

The change will cut the perofrmance impact on 'open' syscall by ~50%, as we now use
a single probe instead of 2 probes.

guyarb committed:
usm, ebpf-platform: Introduce bypassable programs for fexit

A similar macro to BPF_BYPASSABLE_KRETPROBE, with the intent to support
fexit probes. Allows dynamic control of the execution of probes, as in
tests we want to stop the probe and resume it later.

guyarb committed:
usm: sowatcher: Change sysOpenAt2Supported

If the function was missing or we couldn't decude its existence via ddebpf.VerifyKernelFuncs,
We should not assume it exists just based on the kernel version

guyarb committed:
usm: sowatcher: Added a method to determine if fexit is supported

Added a method to dynamically decide if fexit is supported on the system.
While fexit is supported from 5.5 kernels for x86, and 6.1 for arm, it was
unstable and missing from arm kernels up to 6.4. The best way to determine
the existance of the feature is by trying to load a dummy fexit of the probe
and checks if it works. The method was taken from CWS.

guyarb committed:
usm: sowatcher: Encapsulate probes initialization

guyarb committed:
usm: sowatcher: Determine which probes to enable

Determine if we can and should use the more advanced fexit version of the probes
or if we should rely on the old versions via tracepoints

val06 commented on pkg/ebpf/c/bpf_bypass.h:
```diff
@@ -48,4 +48,19 @@ static __always_inline typeof(name(0)) ____##name(struct pt_regs *ctx, ##args)
 #define BPF_BYPASSABLE_UPROBE(name, args...)  BPF_BYPASSABLE_KPROBE(name, ##args)
 #define BPF_BYPASSABLE_URETPROBE(name, args...)  BPF_BYPASSABLE_KRETPROBE(name, ##args)
 
+#define BPF_BYPASSABLE_PROG(name, args...)						    \
```
can you add a short comment? (also for the other existing macros) to explain them (providing an example where they would be used in 1 short paragraph)

guyarb replied:
For BPF_BYPASSABLE_PROG - done
for the rest - in a separate PR

guyarb replied:
https://github.com/DataDog/datadog-agent/pull/31717

val06 commented on pkg/network/usm/sharedlibraries/ebpf.go:
```diff
@@ -537,43 +550,83 @@ func (e *EbpfProgram) initPrebuilt() error {
 	}
 
-	res := make([]manager.ProbeIdentificationPair, 0, len(traceTypes)*len(openatProbes))
+	oldProbes := make([]manager.ProbeIdentificationPair, 0, len(traceTypes)*len(openatProbes))
```
i would call it `tpProbes`

guyarb replied:
done

val06 commented on pkg/network/usm/sharedlibraries/ebpf.go:
```diff
@@ -537,43 +550,83 @@ func (e *EbpfProgram) initPrebuilt() error {
+	openat2Supported := sysOpenAt2Supported()
+	isFexitSupported := fexitSupported("do_sys_openat2")
+
+	advancedProbes := []manager.ProbeIdentificationPair{
```
i would call it `tracingProbes` (and maybe add 1 liner comment explaining it is newer fentry/fexit bpf programs)

guyarb replied:
done

val06 commented on pkg/network/usm/sharedlibraries/ebpf.go:
```diff
@@ -105,6 +110,11 @@ type EbpfProgram struct {
 	isInitialized bool
+
+	// enabledProbes is a list of the probes that are enabled for the current system.
+	enabledProbes []manager.ProbeIdentificationPair
+	// disabledProbes is a list of the probes that are disabled for the current system.
+	disabledProbes []manager.ProbeIdentificationPair
```
i saw a similar approach in other places in our code.
do you think it would make more sense to move that logic to the manager somehow.

it is a bit ridiculous for me to have so many ebpf manager wrappers and still duplicate code like this in the actual business logic of our ebpf products

val06 commented on pkg/network/usm/sharedlibraries/ebpf.go:
```diff
@@ -537,43 +550,83 @@ func (e *EbpfProgram) initPrebuilt() error {
+func fexitSupported(funcName string) bool {
+	if features.HaveProgramType(ebpf.Tracing) != nil {
+		return false
 	}
 
-	kversion, err := kernel.HostVersion()
+	spec := &ebpf.ProgramSpec{
+		Type:       ebpf.Tracing,
+		AttachType: ebpf.AttachTraceFExit,
+		AttachTo:   funcName,
+		Instructions: asm.Instructions{
+			asm.LoadImm(asm.R0, 0, asm.DWord),
+			asm.Return(),
+		},
+	}
+	prog, err := ebpf.NewProgramWithOptions(spec, ebpf.ProgramOptions{
+		LogDisabled: true,
+	})
+	if err != nil {
+		return false
+	}
+	defer prog.Close()
 
+	l, err := link.AttachTracing(link.TracingOptions{
+		Program: prog,
+	})
 	if err != nil {
-		log.Error("could not determine the current kernel version. fallback to do_sys_open")
 		return false
 	}
+	defer l.Close()
 
-	return kversion >= kernel.VersionCode(5, 6, 0)
+	return true
 }
```
this is nice, consider (in a separate PR) introduce a "self-test" function (or interface) as part of the base ebpf program struct. this can allow a better (cleaner) troubleshooting and UX when trying to use USM

val06 reviewed (approved):
added few comments/suggestions (take a look before merging :) ). most of them don't need to be part of this PR, in case you decide to implement, so I am approving this PR.

vitkyrka commented on pkg/network/ebpf/c/shared-libraries/probes.h:
```diff
@@ -144,4 +145,13 @@ int tracepoint__syscalls__sys_exit_openat2(exit_sys_ctx *args) {
 }
 
+SEC("fexit/do_sys_openat2")
+int BPF_BYPASSABLE_PROG(do_sys_openat2_exit, int dirfd, const char *pathname, void *how, long ret) {
```
I think this will make the tests added by https://github.com/DataDog/datadog-agent/pull/32402 to fail since this prog doesn't check for the flags? In that PR, open(2) and openat(2) were supported while openat2(2) was not, but after this PR, all three system calls are handled via this hook point on relevant kernels.

guyarb replied:
added support to ignore write-only calls

guyarb committed:
Merge branch 'main' into guy.arbitman/sowatcher-fentry

guyarb committed:
usm: sowatcher: Rename variables

guyarb committed:
ebpf platform: Added documentation for BPF_BYPASSABLE_PROG

guyarb committed:
usm: sowatcher: Filter write-only calls for openat2

guyarb committed:
usm: sowatcher: Fix reading of flags in openat2 tracepoint

The pointer is a user mode pointer, and we must resolve the value by using bpf_probe_read_user

guyarb committed:
usm: sowatcher: tests: Cover all options for syscalls

The test expanding TestSharedLibraryIgnoreWrite to test directly each syscall type (open, openat, openat2)
instead of relying on glibc support.

guyarb committed:
usm: sowatcher: tests: Skip openat2 on unsupported kernels

vitkyrka commented on pkg/network/usm/sharedlibraries/watcher_test.go:
```diff
@@ -104,40 +106,89 @@ func (s *SharedLibrarySuite) TestSharedLibraryDetection() {
+				fd, err := open(unix.AT_FDCWD, readPath, &how, tt.syscallType)
+				require.NoError(c, err)
+				require.NoError(c, syscall.Close(fd))
+				assert.Equal(c, 1, registerRecorder.CallsForPathID(readPathID))
```
Shouldn't this be `assert.GreaterOrEqual`? If the registration is delayed slightly in the first try you could end up with the `CallsForPathID` returning zero in the first try and 2 in the second one.

vitkyrka commented on pkg/network/usm/sharedlibraries/watcher_test.go:
```diff
@@ -104,40 +106,89 @@ func (s *SharedLibrarySuite) TestSharedLibraryDetection() {
+			}, time.Second*5, 100*time.Millisecond)
+
+			require.EventuallyWithT(t, func(c *assert.CollectT) {
+				how.Flags = syscall.O_CREAT | syscall.O_WRONLY
```
Maybe this `EventuallyWithT` could be combined with the one above to reduce the risk of it succeeding due to the registration being slow? The one above could do write + read + write and then assert for the expected calls.

guyarb committed:
usm: sowatcher: tests: Combine test cases

guyarb commented:
/merge
